api/applications_api.py

```diff
@@ -0,0 +1,188 @@
+
+    def perform_create(self, serializer):
+        serializer.save()
+
+
+class ApplicationSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Application
+        fields = [
+            "id",
+            "description",
+            "reservation_purpose",
+            "organisation",
+            "contact_person",
+            "user",
+            "organisation",
+        ]
+
+    def validate(self, data):
+        return data
+
+
+class ApplicationViewSet(viewsets.ModelViewSet):
+
+    queryset = Application.objects.all()
+
+    serializer_class = ApplicationSerializer
+
+    def perform_create(self, serializer):
+        serializer.save(user=self.request.user)
+
+
+class RecurrenceReadSerializer(serializers.ModelSerializer):
+    def to_representation(self, data):
+        recurrence = {}
+        for field in self.fields:
+            recurrence[field] = getattr(data, field)
+
+        recurrence["recurrence"] = data.recurrence.between(
+            self.get_start_date(),
+            self.get_end_date(),
+        )
+        return recurrence
+
+    def get_start_date(self) -> datetime:
+        if "start" in self.context and self.context["start"] is not None:
+            return self.context["start"]
+        return MINIMUM_TIME
+
+    def get_end_date(self) -> datetime:
+        if "end" in self.context and self.context["end"] is not None:
+            return self.context["end"]
+        return MAXIMUM_TIME
+
+    class Meta:
+        model = Recurrence
+        fields = ["id", "priority"]
+
+
+class ApplicationEventSerializer(serializers.ModelSerializer):
+
+    recurrences = RecurrenceReadSerializer(many=True, read_only=True)
+
+    class Meta:
+        model = ApplicationEvent
+        fields = [
+            "id",
+            "recurrences",
+            "num_persons",
+            "age_group",
+            "ability_group",
+            "num_events",
+            "duration",
+            "application",
+        ]
+
+
+class ApplicationEventViewSet(viewsets.ModelViewSet):
+    def list(self, request):
+
+        queryset = ApplicationEvent.objects.all()
+
+        serializer = ApplicationEventSerializer(
+            queryset, many=True, context=self.get_request_params()
+        )
+
+        return Response(serializer.data)
+
+    def get_request_params(self):
+        start: datetime = self.request.query_params.get("start", None)
+        end: datetime = self.request.query_params.get("end", None)
+
+        if start is not None:
+            start = parse(start)
+
+        if end is not None:
+            end = parse(end)
+
+        return {"start": start, "end": end, "read": self.is_read()}
+
+    def retrieve(self, request, pk=None):
+        queryset = ApplicationEvent.objects.all()
+        user = get_object_or_404(queryset, pk=pk)
+        serializer = ApplicationEventSerializer(user, context=self.get_request_params())
+        return Response(serializer.data)
```

Comment:
Is this only for passing context? If so, it's better to override only the `get_serializer_context` method of `ModelViewSet`.

I'll also note that unless there's a good reason not to, you should always get queryset with `self.get_queryset()` or get the object with `self.get_object()` etc.

http://cdrf.co helps a lot as well.

Reply:
👍